File: tests/drivers/go/v5/read_routing.go
package main

import (
	"fmt"
	"github.com/neo4j/neo4j-go-driver/v5/neo4j"
)

func read_messages(uri string) {
	username := ""
	password := ""

	// Connect to Memgraph
	driver, err := neo4j.NewDriver(uri, neo4j.BasicAuth(username, password, ""))
	if err != nil {
		panic(err)
	}
	defer driver.Close()

	// Use AccessModeRead for read transactions
	session := driver.NewSession(neo4j.SessionConfig{AccessMode: neo4j.AccessModeRead})
	defer session.Close()

	greeting, err := session.ReadTransaction(func(transaction neo4j.Transaction) (interface{}, error) {
		result, err := transaction.Run("MATCH (n:Greeting) RETURN n.message AS message LIMIT 1", nil)
		if err != nil {
			return nil, err
		}

		if result.Next() {
			return result.Record().Values[0], nil
		}

		return nil, result.Err()
	})

	if err != nil {
		panic(err)
	}

	fmt.Println(greeting)

}

// Test checks that you can use bolt+routing for connecting to main and coordinators for reading.
func main() {
	fmt.Println("Started running read_route.go test")
	read_messages("neo4j://localhost:7690") // coordinator_1
	read_messages("neo4j://localhost:7691") // coordinator_2
	read_messages("neo4j://localhost:7692") // coordinator_3
	fmt.Println("Successfully finished running read_route.go test")
}

File: tests/drivers/java/v5_8/src/main/java/memgraph/ReadRouting.java
package memgraph;

import static org.neo4j.driver.Values.parameters;

import java.util.*;
import java.util.concurrent.TimeUnit;
import org.neo4j.driver.AuthTokens;
import org.neo4j.driver.Driver;
import org.neo4j.driver.GraphDatabase;
import org.neo4j.driver.Session;
import org.neo4j.driver.Transaction;

public class ReadRouting {
    private Driver driver;

    private void readMessage(String uri) {
        driver = GraphDatabase.driver(uri, AuthTokens.basic("", ""));
        try (Session session = driver.session()) {
            String greeting = session.readTransaction(tx -> {
                var result = tx.run("MATCH (n:Greeting) RETURN n.message AS message");
                System.out.println("Read txn passed!");
                return "OK";
            });
        }
    }

    public static void main(String... args) {
        System.out.println("Started running ReadRoutingTest...");
        ReadRouting greeter = new ReadRouting();
        greeter.readMessage("neo4j://localhost:7690"); // coordinator_1
        greeter.readMessage("neo4j://localhost:7691"); // coordinator_2
        greeter.readMessage("neo4j://localhost:7692"); // coordinator_3
        System.out.println("All good!");
    }
}
